res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,718 @@
+
+PioneerDDJ400.keyshiftModePad = function(_channel, control, value, _status, group){
+    if(value == 0) return; // ignore release
+    engine.setValue(group, 'pitch', this.halftoneToPadMap[control & 0xf]);
+};
+
+PioneerDDJ400.samplerModeShiftPadPressed = function(_channel, _control, value, _status, group){
+    if(value == 0) {
+        return; // ignore release
+    }
+    var playing = engine.getValue(group, 'play');
+    // when playing stop and return to start/cue point
+    if(playing > 0){
+        engine.setValue(group, 'cue_gotoandstop', 1);
+    }
+    else{ // load selected track
+        engine.setValue(group, 'LoadSelectedTrack', 1);
+    }
+    // TODO: while playing a sample blink playing PAD?
+};
+
+PioneerDDJ400.beatjumpPadPressed = function(_channel, control, value, status, group){
+    if(value == 0) {
+        midi.sendShortMsg(status, control, 0); // turn off LED
+        return; // ignore release
+    }
+
+    const newVal = this.beatjumpPad[control & 0xf] * this.beatjumpMulitplier;
+
+    engine.setValue(group, 'beatjump', newVal);
+    midi.sendShortMsg(status, control, 0x7f); // turn LED back on
+};
+
+PioneerDDJ400.beatjumpShiftByOne = function(_channel, control, value, status){
+    if(value == 0) {
+        midi.sendShortMsg(status, control, 0); // turn off LED
+        return; // ignore release
+    }
+    
+    var direction = status <= 0x98 ? -1 : 1;
+    
+    if ( direction == -1 && this.beatjumpMulitplier <= 1){
+        direction = 0;
+    }
+
+    this.beatjumpMulitplier += direction;
+    midi.sendShortMsg(status, control, 0x7f); // turn LED back on
+};
+
+
+PioneerDDJ400.shiftPressed = function(channel, _control, value){
+    this.shiftState[channel] = value;
+};
+
+
+PioneerDDJ400.hotcuePadPressed = function(_channel, control, value, status, group){
+    const loopPlaying = engine.getValue(group, 'loop_enabled');
+    const padNum = (control & 0xf) +1;
+    const loopPoint = PioneerDDJ400.hotcueLoopPoints[group][padNum-1];
+    const hotcueSet = engine.getValue(group, 'hotcue_'+padNum+'_enabled');
+
+    // play hotcue if set or set if not set and no loop is playing
+    if(hotcueSet > 0 || (loopPlaying == 0 && (!loopPoint || loopPoint.start <= -1))){
+        engine.setValue(group, 'hotcue_'+padNum+'_activate', value);
+        midi.sendShortMsg(status, control, 0x7f); // enable PAD LED
+        return;
+    }
+
+    // play saved loop if set
+    if(value > 0 && loopPoint && loopPoint.start >= 0){
+        engine.setValue(group, 'loop_start_position', loopPoint.start);
+        engine.setValue(group, 'loop_end_position', loopPoint.end);
+        engine.setValue(group, 'loop_in_goto', 1);
+        engine.setValue(group, 'reloop_toggle', 1);
+        // This is a hack, TODO find a command like loop_in_goto_and_play
+        if(engine.getValue(group, 'loop_enabled') == 0){
+            engine.setValue(group, 'reloop_toggle', 1);
+        }
+        return;
+    }
+
+    // save playing loop on hotcue pad
```

Comment:
I think this hack should be removed. What is the point of saving a loop if it is lost when Mixxx shuts down? We are currently working on adding this feature to Mixxx in #2194. If you would like to help implement and/or test that feature, that would be great!